File: src/main/java/algo/slidingWindow/PermutationInString_567.java
package algo.slidingWindow;

public class PermutationInString_567 {
  public static boolean checkInclusion(String s1, String s2) {
    if (s1.length() > s2.length()) {
      return false;
    }

    var window = s1.length();
    var freq = new int[26];
    for (var c : s1.toCharArray()) {
      freq[c - 'a']++;
    }

    var count = new int[26];
    for (var i = 0; i < window; i++) {
      count[s2.charAt(i) - 'a']++;
    }

    for (var i = 0; i <= s2.length() - window; i++) {
      if (i > 0) {
        count[s2.charAt(i + window - 1) - 'a']++;
        count[s2.charAt(i - 1) - 'a']--;
      }
      if (match(count, freq)) {
        return true;
      }
    }
    return false;
  }

  private static boolean match(int[] count, int[] freq) {
    for (var i = 0; i < count.length; i++) {
      if (count[i] != freq[i]) {
        return false;
      }
    }
    return true;
  }

  public static void main(String[] args) {
    System.out.println(checkInclusion("ab", "eidbaooo"));
    System.out.println(checkInclusion("ab", "eidboaoo"));
    System.out.println(checkInclusion("adc", "dcda"));
  }
}

File: src/main/kotlin/algo/slidingWindow/PermutationInString_567.kt
package algo.slidingWindow

import algo.backtrack.KtPermutations_II_47

class KtPermutationInString_567 {
    fun checkInclusion(s1: String, s2: String): Boolean {
        if (s2.length < s1.length) {
            return false
        }
        val base = IntArray(26)
        s1.forEach { base[it - 'a']++ }

        val count = IntArray(26)
        for (i in s1.indices) {
            count[s2[i] - 'a']++
        }
        for (i in s1.length - 1..<s2.length) {
            if (i > s1.length - 1) {
                count[s2[i] - 'a']++
                count[s2[i - s1.length] - 'a']--
            }
            // check permutation
            var isPermutation = true
            for (k in 0..25) {
                if (base[k] != count[k]) {
                    isPermutation = false
                    break
                }
            }
            if (isPermutation) {
                return true
            }
        }
        return false
    }
}

fun main() {
    val permutation = KtPermutationInString_567()
    println(permutation.checkInclusion("ab", "eidbaooo"))
}